File: driver-reactive-streams/src/test/functional/com/mongodb/reactivestreams/client/TestSubscriber.java
/*
 * Copyright 2008-present MongoDB, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mongodb.reactivestreams.client;

import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;
import java.util.function.Supplier;

import static com.mongodb.ClusterFixture.TIMEOUT_DURATION;

public class TestSubscriber<T> implements Subscriber<T> {
    private final CountDownLatch latch = new CountDownLatch(1);
    private final ArrayList<T> onNextEvents = new ArrayList<>();
    private final ArrayList<Throwable> onErrorEvents = new ArrayList<>();
    private final ArrayList<Void> onCompleteEvents = new ArrayList<>();

    private Consumer<Subscription> doOnSubscribe = sub -> {};
    private Consumer<T> doOnNext = r -> {};

    private Subscription subscription;

    public TestSubscriber() {
    }

    @Override
    public void onSubscribe(final Subscription subscription) {
        this.subscription = subscription;
        doOnSubscribe.accept(subscription);
    }

    public void doOnSubscribe(final Consumer<Subscription> doOnSubscribe) {
        this.doOnSubscribe = doOnSubscribe;
    }

    public void doOnNext(final Consumer<T> doOnNext) {
        this.doOnNext = doOnNext;
    }

    /**
     * Provides the Subscriber with a new item to observe.
     * <p>
     * The {@code Publisher} may call this method 0 or more times.
     * </p>
     * <p>
     * The {@code Publisher} will not call this method again after it calls either {@link #onComplete} or
     * {@link #onError}.
     * </p>
     * @param result the item emitted by the obserable
     */
    @Override
    public void onNext(final T result) {
        doOnNext.accept(result);
        onNextEvents.add(result);
    }

    /**
     * Notifies the Subscriber that the obserable has experienced an error condition.
     * <p>
     * If the obserable calls this method, it will not thereafter call {@link #onNext} or
     * {@link #onComplete}.
     * </p>
     *
     * @param e the exception encountered by the obserable
     */
    @Override
    public void onError(final Throwable e) {
        try {
            onErrorEvents.add(e);
        } finally {
            latch.countDown();
        }
    }

    /**
     * Notifies the Subscriber that the obserable has finished sending push-based notifications.
     * <p>
     * The obserable will not call this method if it calls {@link #onError}.
     * </p>
     */
    @Override
    public void onComplete() {
        try {
            onCompleteEvents.add(null);
        } finally {
            latch.countDown();
        }
    }

    /**
     * Allow calling the protected {@link Subscription#request(long)} from unit tests.
     *
     * @param n the maximum number of items you want the obserable to emit to the Subscriber at this time, or
     *          {@code Long.MAX_VALUE} if you want the obserable to emit items at its own pace
     */
    public void requestMore(final long n) {
        subscription.request(n);
    }

    /**
     * Get the {@link Throwable}s this {@code Subscriber} was notified of via {@link #onError} as a
     * {@link List}.
     *
     * @return a list of the Throwables that were passed to this Subscriber's {@link #onError} method
     */
    public List<Throwable> getOnErrorEvents() {
        return onErrorEvents;
    }

    /**
     * Get the sequence of items observed by this {@link Subscriber}, as an ordered {@link List}.
     *
     * @return a list of items observed by this Subscriber, in the order in which they were observed
     */
    public List<T> getOnNextEvents() {
        return onNextEvents;
    }

    /**
     * Assert that a particular sequence of items was received by this {@link Subscriber} in order.
     *
     * @param items the sequence of items expected to have been observed
     * @throws AssertionError if the sequence of items observed does not exactly match {@code items}
     */
    public void assertReceivedOnNext(final List<T> items) {
        if (!waitFor(() -> getOnNextEvents().size() == items.size())) {
            throw new AssertionError("Number of items does not match. Provided: " + items.size() + "  Actual: " + getOnNextEvents().size());
        }

        for (int i = 0; i < items.size(); i++) {
            if (items.get(i) == null) {
                // check for null equality
                if (onNextEvents.get(i) != null) {
                    throw new AssertionError("Value at index: " + i + " expected to be [null] but was: [" + getOnNextEvents().get(i) + "]");
                }
            } else if (!items.get(i).equals(getOnNextEvents().get(i))) {
                throw new AssertionError("Value at index: " + i + " expected to be ["
                        + items.get(i) + "] (" + items.get(i).getClass().getSimpleName() + ") but was: [" + getOnNextEvents().get(i)
                        + "] (" + getOnNextEvents().get(i).getClass().getSimpleName() + ")");

            }
        }
    }

    /**
     * Assert that a single terminal event occurred, either {@link #onComplete} or {@link #onError}.
     *
     * @throws AssertionError if not exactly one terminal event notification was received
     */
    public void assertTerminalEvent() {
        try {
            //noinspection ResultOfMethodCallIgnored
            latch.await(TIMEOUT_DURATION.toMillis(), TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }

        if (onErrorEvents.size() > 1) {
            throw new AssertionError("Too many onError events: " + onErrorEvents.size());
        }

        if (onCompleteEvents.size() > 1) {
            throw new AssertionError("Too many onCompleted events: " + onCompleteEvents.size());
        }

        if (onCompleteEvents.size() == 1 && onErrorEvents.size() == 1) {
            throw new AssertionError("Received both an onError and onCompleted. Should be one or the other.");
        }

        if (onCompleteEvents.size() == 0 && onErrorEvents.size() == 0) {
            throw new AssertionError("No terminal events received.");
        }
    }

    /**
     * Assert that no terminal event occurred, either {@link #onComplete} or {@link #onError}.
     *
     * @throws AssertionError if a terminal event notification was received
     */
    public void assertNoTerminalEvent() {
        if (onCompleteEvents.size() != 0 && onErrorEvents.size() != 0) {
            throw new AssertionError("Terminal events received.");
        }
    }

    /**
     * Assert that this {@link Subscriber} has received no {@code onError} notifications.
     *
     * @throws AssertionError if this {@link Subscriber} has received one or more {@link #onError} notifications
     */
    public void assertNoErrors() {
        if (onErrorEvents.size() > 0) {
            throw new AssertionError("Unexpected onError events: " + getOnErrorEvents().size(), getOnErrorEvents().get(0));
        }
    }

    private boolean waitFor(final Supplier<Boolean> check) {
        int retry = 0;
        long totalSleepTimeMS = 0;
        while (totalSleepTimeMS < TIMEOUT_DURATION.toMillis()) {
            retry++;
            if (check.get()) {
                return true;
            }
            long sleepTimeMS = 100 + (100 * (long) Math.pow(retry, 2));
            totalSleepTimeMS += sleepTimeMS;
            try {
                Thread.sleep(sleepTimeMS);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return false;
            }
        }
        return false;
    }

}

File: driver-scala-old/examples/src/test/scala/reactivestreams/TestSubscriber.scala
/*
 * Copyright 2008-present MongoDB, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package reactivestreams

import java.util.concurrent.{ CountDownLatch, TimeUnit }

import org.mongodb.scala.Completed

import scala.collection.mutable.ArrayBuffer
import scala.concurrent.duration.Duration
import scala.util.{ Failure, Success, Try }
import org.reactivestreams.{ Subscriber, Subscription }

object TestSubscriber {
  def apply[T](): TestSubscriber[T] = {
    TestSubscriber(new Subscriber[T]() {
      def onSubscribe(subscription: Subscription) {}

      def onNext(result: T) {}

      def onComplete() {}

      def onError(e: Throwable) {}
    })
  }
}

case class TestSubscriber[T](delegate: Subscriber[T]) extends Subscriber[T] {
  private final val latch: CountDownLatch = new CountDownLatch(1)
  private final val onNextEvents: ArrayBuffer[T] = new ArrayBuffer[T]
  private final val onErrorEvents: ArrayBuffer[Throwable] = new ArrayBuffer[Throwable]
  private final val onCompleteEvents: ArrayBuffer[Completed] = new ArrayBuffer[Completed]

  private var subscription: Option[Subscription] = None

  def onSubscribe(subscription: Subscription) {
    this.subscription = Some(subscription)
  }

  /**
   * Provides the Subscriber with a new item to observe.
   *
   * The `Publisher` may call this method 0 or more times.
   *
   * The `Publisher` will not call this method again after it calls either `onComplete` or `onError`.
   *
   * @param result the item emitted by the obserable
   */
  def onNext(result: T): Unit = {
    onNextEvents += result
    delegate.onNext(result)
  }

  /**
   * Notifies the Subscriber that the obserable has experienced an error condition.
   *
   * If the obserable calls this method, it will not thereafter call `onNext` or
   * `onComplete`.
   *
   * @param e the exception encountered by the obserable
   */
  def onError(e: Throwable): Unit = {
    try {
      onErrorEvents += e
      delegate.onError(e)
    } finally {
      latch.countDown()
    }
  }

  /**
   * Notifies the Subscriber that the obserable has finished sending push-based notifications.
   *
   * The obserable will not call this method if it calls `onError`.
   *
   */
  def onComplete(): Unit = {
    try {
      onCompleteEvents += null
      delegate.onComplete()
    } finally {
      latch.countDown()
    }
  }

  /**
   * Allow calling the protected `Subscription.request(long)` from unit tests.
   *
   * @param n the maximum number of items you want the obserable to emit to the Subscriber at this time, or
   *          `Long.MaxValue` if you want the obserable to emit items at its own pace
   */
  def requestMore(n: Long): Unit = {
    subscription match {
      case Some(sub) => sub.request(n)
      case None =>
    }
  }

  /**
   * Get the `Throwable`s this `Subscriber` was notified of via `onError` as a `Seq`.
   *
   * @return a list of the Throwables that were passed to this Subscriber's { @link #onError} method
   */
  def getOnErrorEvents: Seq[Throwable] = onErrorEvents.toSeq

  /**
   * Get the sequence of items observed by this `Subscriber`, as an ordered `List`.
   *
   * @return a list of items observed by this Subscriber, in the order in which they were observed
   */
  def getOnNextEvents: Seq[T] = onNextEvents.toSeq

  /**
   * Returns the subscription to the this `Subscriber`.
   *
   * @return the subscription or null if not subscribed to
   */
  def getSubscription: Option[Subscription] = subscription

  /**
   * Assert that a particular sequence of items was received by this `Subscriber` in order.
   *
   * @param items the sequence of items expected to have been observed
   * @throws AssertionError if the sequence of items observed does not exactly match `items`
   */
  def assertReceivedOnNext(items: Seq[T]) {
    if (getOnNextEvents.size != items.size) {
      throw new AssertionError(s"Number of items does not match. Provided: ${items.size} Actual: ${getOnNextEvents.size}")
    }

    items.indices.foreach(i => {
      items(i) == onNextEvents(i) match {
        case false => throw new AssertionError(s"Value at index: $i expected to be [${items(i)}] but was: [${onNextEvents(i)}]")
        case true =>
      }
    })
  }

  /**
   * Assert that a single terminal event occurred, either `onComplete` or `onError`.
   *
   * @throws AssertionError if not exactly one terminal event notification was received
   */
  def assertTerminalEvent(): Unit = {
    if (onErrorEvents.size > 1) {
      throw new AssertionError("Too many onError events: " + onErrorEvents.size)
    }
    if (onCompleteEvents.size > 1) {
      throw new AssertionError("Too many onCompleted events: " + onCompleteEvents.size)
    }
    if (onCompleteEvents.size == 1 && onErrorEvents.size == 1) {
      throw new AssertionError("Received both an onError and onCompleted. Should be one or the other.")
    }
    if (onCompleteEvents.isEmpty && onErrorEvents.isEmpty) {
      throw new AssertionError("No terminal events received.")
    }
  }

  /**
   * Assert that no terminal event occurred, either `onComplete` or `onError`.
   *
   * @throws AssertionError if a terminal event notification was received
   */
  def assertNoTerminalEvent(): Unit = {
    if (onCompleteEvents.nonEmpty && onErrorEvents.nonEmpty) {
      throw new AssertionError("Terminal events received.")
    }
  }

  /**
   * Assert that this `Subscriber` has received no `onError` notifications.
   *
   * @throws AssertionError if this { @link Subscriber} has received one or more { @link #onError} notifications
   */
  def assertNoErrors(): Unit = {
    if (onErrorEvents.nonEmpty) {
      throw new RuntimeException("Unexpected onError events: " + onErrorEvents.size, getOnErrorEvents.head)
    }
  }

  /**
   * Assert that this `Subscriber` has received an `onError` notification.
   *
   * @throws AssertionError if this { @link Subscriber} did not received an { @link #onError} notifications
   */
  def assertErrored(): Unit = {
    if (onErrorEvents.isEmpty) {
      throw new RuntimeException("No onError events")
    }
  }

  /**
   * Blocks until this `Subscriber` receives a notification that the `Observable` is complete (either an `onCompleted` or
   * `onError` notification).
   *
   * @throws RuntimeException if the Subscriber is interrupted before the Observable is able to complete
   */
  def awaitTerminalEvent(): Unit = {
    Try(latch.await()) match {
      case Failure(ex) => throw new RuntimeException("Interrupted", ex)
      case _ =>
    }
  }

  /**
   * Blocks until this `Subscriber` receives a notification that the `Observable` is complete (either an `onCompleted` or
   * `onError` notification).
   *
   * @param duration the duration of the timeout
   * @throws RuntimeException
   *          if the Subscriber is interrupted before the Observable is able to complete
   */
  def awaitTerminalEvent(duration: Duration) {
    Try(latch.await(duration.toMillis, TimeUnit.MILLISECONDS)) match {
      case Failure(ex) => throw new RuntimeException("Interrupted", ex)
      case Success(false) => throw new RuntimeException("Failed to return in time")
      case Success(true) =>
    }
  }
}

